### PR TITLE
Release single-step evaluation framework and wrappers for several model types

### PR DESCRIPTION
**Pull request by kmaziarz**
This (fairly large) PR releases the majority of the single-step evaluation code, including the eval script itself, inference wrappers for seven model types, and glue code to plug these models into search. The changes are split into semantically meaningful commits. Some notes for reviewing:
- Wherever the commit message starts with "Release", this means our internal code was copied verbatim (or almost verbatim), and thus at the very least it went through code review in the past.
- Some of the commits edit the configuration around CI and pre-commit. These changes were needed partly because the model wrappers contain dependencies which are not installed in our default environment, and thus one needs to tell `mypy` to ignore external libraries it doesn't know about.
- Some bits of internal code are not part of this release; the  "Make the evaluation pipeline easy to extend" commit makes it easier to plug them in into `syntheseus` from outside.

A few notable things which are still missing, to be added in future PRs to avoid making this one even larger:
- Detailed setup instructions for each model type (e.g. environment definitions).
- Model checkpoints for each model type.
- More tests covering `syntheseus/reaction_prediction` (currently the coverage is quite low, partly because the wrappers themselves cannot be tested easily without setting up dedicated environments for each, but there are also places which are just not adequately covered by tests yet).

### CODE REVIEW

**Review by AustinT — APPROVED** [review]
I left a few small comments, but overall looks good to me. The tests all pass locally for me.

I was not able to run `eval.py` locally because I don't have the correct model files or example configs, so I can't independently confirm that it works. Did you test it locally @kmaziarz ? If so I am happy to merge it, and provide documentation / configs / model weights in a future PR.

**Comment by kmaziarz** [comment]
> I was not able to run eval.py locally because I don't have the correct model files or example configs, so I can't independently confirm that it works. By extension, I could not test the "Make the evaluation pipeline easy to extend" commit. Did you test it locally @kmaziarz?

Yes, I re-ran the single-step evaluation for all seven model types, and confirmed that all resulting numbers match the table in the paper 🙂

File: pyproject.toml [hunk 1]
```diff
@@ -52,8 +52,9 @@ namespaces = false
 line-length = 100
 include = '\.pyi?$'
 
-[tool.mypy.overrides]
+[tool.mypy]
```
[the remark this answers is not in the document]
Comment (kmaziarz) [reply pyproject.toml:55]: I'm not sure how the old `[tool.mypy.overrides]` section was even interpreted; my understanding is that you'd either use `[[tool.mypy.overrides]]` and list overrides per module, or `[tool.mypy]` and list global ones...

File: syntheseus/reaction_prediction/chem/utils.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+def molecule_bag_from_smiles_strict(smiles: str) -> Bag[Molecule]:
+    return Bag([Molecule(component) for component in smiles.split(SMILES_SEPARATOR)])
+
+
+def molecule_bag_from_smiles(smiles: str) -> Optional[Bag[Molecule]]:
+    try:
+        return molecule_bag_from_smiles_strict(smiles)
+    except ValueError:
+        # If any of the components ends up invalid we return `None` instead.
+        return None
```
[the remark this answers is not in the document]
Comment (kmaziarz) [reply syntheseus/reaction_prediction/chem/utils.py:41]: The reason this is split is because in the unreleased code we do sometimes use `molecule_bag_from_smiles_strict` directly (in cases when we don't expect to get invalid things). Of course in most cases I tried to make the release coherent by not releasing e.g. utils that are only used in the unreleased code, but here it seems harmless, because `molecule_bag_from_smiles_strict` _is_ used, just in a single place as you note.

File: syntheseus/reaction_prediction/data/dataset.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    def get_num_samples(self, fold: DataFold) -> int:
+        pass
+
+    @classmethod
+    def get_data_path(cls, data_dir: Union[str, Path], fold: DataFold) -> Path:
+        return Path(data_dir) / f"{fold.value}.jsonl"
+
+    @classmethod
+    def sample_from_json(cls, data: str, sample_cls: Type[SampleType]) -> SampleType:
+        return sample_cls.from_dict(json.loads(data))
```
[the remark this answers is not in the document]
Comment (kmaziarz) [reply syntheseus/reaction_prediction/data/dataset.py:39]: Yes, each fold is a single plain `jsonl` file, but for training we would use a more compressed representation of the featurized data (which does allow for e.g. random access). The purpose of this `ReactionDataset` class is twofold: before training it provides raw data which goes into a featurization function, and during evaluation it feeds directly into the inference wrappers. In both cases, it essentially has to yield all samples, just once, and in order (no shuffling). It does so in a streaming manner, so the data never has to be fully instantiated in memory. The only problem could be if the file containing the raw data is just too big for whatever system-related reason, but we never found that to be an issue. We used to use a compressed `jsonl` format here, but if we don't need the compression then it's nice to skip it, as then one can e.g. grep things in the data file directly.

[the remark this answers is not in the document]
Comment (AustinT) [reply syntheseus/reaction_prediction/data/dataset.py:39]: Sounds good, I'll mark this as resolved. I just raised this point in case it was something we never considered.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -1,3 +1,6 @@
+default_language_version:
+  python: python3
```
[the remark this answers is not in the document]
Comment (AustinT) [reply .pre-commit-config.yaml:2]: Makes sense. You should update your conda though 😝

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -40,6 +40,7 @@ repos:
       - id: mypy
         name: "mypy"
         files: "syntheseus/"
+        args: ["--install-types", "--non-interactive"]
```
[the remark this answers is not in the document]
Comment (kmaziarz) [reply .pre-commit-config.yaml:43]: For some external libraries, the typing is not provided in the library itself, but in separate type stub libraries that can be installed; these options tell `mypy` to automatically install these type stubs into its environment when doing the type checking. On the other hand, given that we set `ignore_missing_imports = true`, normally `mypy` would just not worry about these untyped libraries. However, there's one library which (due to some historical reasons) `mypy` doesn't want to ignore and really wants to install the type stubs for. By combining `ignore_missing_imports = true` with `--install-types`, `mypy` will install some typing stubs where available and ignore those it doesn't have stubs for.

(As a side note, we do use `ignore_missing_imports` when running `mypy` on the internal codebase; it seems it eventually becomes necessary once one has enough dependencies, as some libraries just don't expose any types and you have to tell `mypy` to not worry about them)

File: syntheseus/tests/reaction_prediction/chem/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    smiles_mapped = "[OH:1][CH2:2][c:3]1[cH:4][n:5][cH:6][cH:7][c:8]1[Br:9]"
+    smiles_unmapped = "OCc1cnccc1Br"
+
+    assert remove_atom_mapping(smiles_mapped) == smiles_unmapped
+    assert remove_atom_mapping(smiles_unmapped) == smiles_unmapped
+
+    mol = Chem.MolFromSmiles(smiles_mapped)
+    remove_atom_mapping_from_mol(mol)
+
+    assert Chem.MolToSmiles(mol) == smiles_unmapped
```
[the remark this answers is not in the document]
Comment (kmaziarz) [reply syntheseus/tests/reaction_prediction/chem/test_utils.py:19]: If we're talking about this specific test, then there's not much that can happen here; removing the atom mapping is just a matter of clearing a certain atom property, it doesn't really matter how the molecule looks like or if the mapping was at all valid. The `remove_atom_mapping*` functions don't claim to return a _canonical_ molecule either, they just return something without the mapping, and canonicalization is handled elsewhere.

File: pyproject.toml [hunk 7]
```diff
@@ -16,9 +16,13 @@ requires-python = ">=3.7"
...
-    "numpy",
-    "rdkit",
-    "networkx",
+    "more_itertools",       # reaction_prediction
+    "networkx",             # search
+    "numpy",                # reaction_prediction, search
+    "omegaconf",            # reaction_prediction
+    "pydantic>=1.10.5,<2",  # reaction_prediction
+    "rdkit",                # reaction_prediction, search
+    "tqdm",                 # reaction_prediction
```
[the remark this answers is not in the document]
Comment (kmaziarz) [reply pyproject.toml:25]: Good point, I tried to update `README.md` to mention the two subpackages and how a minimalistic install (e.g. search-only) could work. There will be another update needed once we add all the single-step model environments and checkpoints, but it's a start. Take a look!

File: syntheseus/tests/reaction_prediction/chem/test_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (AustinT) [reply syntheseus/tests/reaction_prediction/chem/test_utils.py:19]: I was just wondering if there are any edge cases where it can go wrong: for example, where the unmapping version will could accidentally alter the graph structure of the molecule, change chirality, etc. If the function is basically just removing numbers and you think nothing can go wrong then I think the test is enough.

File: pyproject.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (AustinT) [reply pyproject.toml:25]: Looks good to me!
